### ISSUE TITLE
Error "(-215:Assertion failed) (int)_numAxes == inputs[0].size() in function 'cv::dnn::PermuteLayerImpl::getMemoryShapes" during net.forward for blob containng multiple images

### ISSUE THREAD

**Issue by TrueWodzu** [opencv/opencv#22904]
### System Information

OpenCV python version: 4.6.0 (built from c++ sources with cuda)
Operating System / Platform: Windows 10
Python version: 3.10

### Detailed description

Hello,

I am trying to do detections on multiple images at once. So I am putting them into a blob by using `cv2.dnn.blobFromImages` but during net.forward I am getting error:
`cv2.error: OpenCV(4.6.0) D:\CODING\Developement\C++\Libs\OpenCV 4.6.0\modules\dnn\src\layers\permute_layer.cpp:161: error: (-215:Assertion failed) (int)_numAxes == inputs[0].size() in function 'cv::dnn::PermuteLayerImpl::getMemoryShapes'`

Detecting works fine for single image and function `cv2.dnn.blobFromImage`

Model has been converted to onnx from YOLOv5 .pt

[Data.zip](https://github.com/opencv/opencv/files/10139199/Data.zip)

### Steps to reproduce

```
INPUT_WIDTH = 640
INPUT_HEIGHT = 640	

modelWeights = "yolov5n.onnx"
net = cv2.dnn.readNet(modelWeights)

images = []
images.append(cv2.imread("1.jpg"))
images.append(cv2.imread("2.jpg"))

blob = cv2.dnn.blobFromImages(images, 1/255, (INPUT_WIDTH, INPUT_HEIGHT), [0,0,0], 1, crop=False)
net.setInput(blob)
output_layers = net.getUnconnectedOutLayersNames()
outputs = net.forward(output_layers)
return outputs
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by TrueWodzu** [comment]
Someone has the same problem here: https://stackoverflow.com/questions/73996296/batch-inference-for-onnx-opencv-c

**Comment by TrueWodzu** [comment]
Not a bug. I've passed onnx model that expects only one image as an input.

**Comment by zihaomu** [comment]
> Not a bug. I've passed onnx model that expects only one image as an input.
Yes, the model expects the shape is [1,3,640,640] while the blob of two image is [2, 3, 640, 640].

**Comment by TrueWodzu** [comment]
@zihaomu any chance you could look at the problem #19347 we really need a support for variable input in model.

**Comment by zihaomu** [comment]
Hi @TrueWodzu, your model has frozen the input shape as [1,3,640,640]. And if you wanna the muti-batch input, you can try to re-export the input shape as `[?, 3, 640, 640]`. 
Some useful answers: https://stackoverflow.com/questions/67718905/change-input-size-of-onnx-model.
The reason why the current model cannot be used directly for multi-batch is that it contains the reshape layer and the reshape layer expects the output shape is [ 1,3,85,80,80] which is only for batch size 1. (layer name `Reshape_199`). Re-export can handle this issue.

**Comment by TrueWodzu** [comment]
Hi @zihaomu I know abou that, after all I wrote that myself. But then, I've exported the model with dynamic input shape and I've asked you to look here: https://github.com/opencv/opencv/issues/19347 There I am presenting other error.

**Comment by zihaomu** [comment]
Hi @TrueWodzu, Thanks for your details. Can you attach your batch-dynamic model in this issue?

There are two different dynamic shape: batch size dynamic shape (could be supported in OpenCV) input shape like `[?, 3, 640, 640]`, fully dynamic shape (could not be supported.), the input shape like `[1, 3, ?, ?]` or `[batch, 3, w, h]`.

For a fully dynamic shape model, some models that do not contain dimensionally sensitive layers (like `reshape`, and `transpose`) **may be supported**. 
In conclusion, currently we do not recommend users to use OpenCV for inference with dynamic input models. And full dynamic shape will be supported in a next 5.x branch, which is in development.

**Comment by TrueWodzu** [comment]
Hi @zihaomu  many thanks for your attention to this. I've attached the model in https://github.com/opencv/opencv/issues/19347#issuecomment-1337017510 but for your convenience I will attach it [here ](https://drive.google.com/file/d/1E7dzTI-vpUDSeQDsXMASM4q1931Eg55h/view?usp=sharing )as well.

I was able to create a pre-determined batch size model, for example [4, 3, w, h]. and run it in OpenCV. It is good for start but as you can imagine, sometimes you want to pass 3 images, sometimes 7, if you have them for example in a queue.

Will wait for branch 5.x Thank you for your work.

**Comment by dkurt** [comment]
> It is good for start but as you can imagine, sometimes you want to pass 3 images, sometimes 7 etc., for example if you have them in a queue ready for processing.

Unfortunately, that would be inefficient in OpenCV because of re-initialization on every new batch size. So try choose a fixed batch size and fill mussed data by zeros, for example.

YOLOv5n model test has been added, but only for a specific batch size: https://github.com/opencv/opencv/pull/24553

**Comment by dkurt** [comment]
@TrueWodzu, you still can process remaining 3 images but just add a dummy image with zero pixels. So this is still a batch of 4 but last sample does not consider in final prediction.
